docs/serialization.md

```diff
@@ -0,0 +1,71 @@
+# Serialization
+
+Dark has a lot of serializers. At time of writing:
+
+- binary serializers to store user code
+- many, many dval serializers to show code to users, save code in the DB, respond to
+  API requests, etc. These are mostly handwritten formats (and are often buggy as a result).
+- the "Vanilla" Json serializer which serializes arbitrary types
+- the "OCamlCompatible" Json serializer which is used in situations where it needs to
+  be compatible with the old OCaml json serializers - we would like to replace this
+  with the Vanilla serializer in all cases
+
+## Goals
+
+Our goals with serializers are:
+
+- never change a serialization format in a way that could break users' code
+- deprecate and remove old/broken serialization formats
+- remove the OCamlCompatible serializer
+
+## Testing
+
+To accomplish these, we have robust serialization tests, implemented in Serialization.Tests.fs
+
+For each serialization format in Dark, we store serialized data from that format, to
+ensure it never changes. The serialized data should be quite extensive, covering as
+many cases as we can reasonably think of.
+
+They help solve our goals in the following ways.
+
+### Goal: never changing formats
+
+If the format changes, we want to know about it. The saved serialized data ensures this: if we change something, there's a git diff to tell us something changed, and we can choose to commit it if that's a backward-compatible change or we determine that it's otherwise OK.
```

Comment:
only place in this file where we're not following the rule of "keep lines kinda short"